File: 2022/examples/puzzle_11/main.py
from collections import defaultdict
from functools import reduce


def parse_monkeys(lines: str) -> dict:
    monkeys = {}
    for line in lines:
        if not line:
            monkeys[monkey_index] = dict(
                items=items,
                op=op,
                test=test,
                true_index=true_index,
                false_index=false_index,
            )
        elif "Monkey" in line:
            monkey_index = int(line.rstrip(":").split(" ")[-1])
        elif "Starting" in line:
            _, raw_items = line.split("items: ")
            items = [int(i) for i in raw_items.split(",")]
        elif "Operation" in line:
            _, raw_op = line.split("new =")
            if "old * old" in raw_op:
                op = ("square",)
            elif "*" in raw_op:
                multiplier = int(line.split("* ")[-1])
                op = ("*", multiplier)
            elif "+" in raw_op:
                adder = int(line.split("+ ")[-1])
                op = ("+", adder)
            elif "-" in raw_op:
                minuser = int(line.split("- ")[-1])
                op = ("-", minuser)
            else:
                raise ValueError
        elif "Test" in line:
            test = int(line.split(" ")[-1])
        elif "true" in line:
            true_index = int(line.split(" ")[-1])
        elif "false" in line:
            false_index = int(line.split(" ")[-1])

    monkeys[monkey_index] = dict(
        items=items, op=op, test=test, true_index=true_index, false_index=false_index
    )
    return monkeys


def process_monkeys(monkeys: dict, num_rounds: int = 20, divide_by_3: bool = False):
    num_items_per_monkey = defaultdict(int)
    max_modulo = reduce(
        lambda a, b: a * b, [monkey["test"] for monkey in monkeys.values()]
    )
    for _ in range(1, num_rounds + 1):
        for monkey_index in monkeys:
            monkey = monkeys[monkey_index]
            num_items_per_monkey[monkey_index] += len(monkey["items"])
            for item in monkey["items"]:
                if monkey["op"] == ("square",):
                    new_item = item * item
                else:
                    op, value = monkey["op"]
                    if op == "*":
                        new_item = item * value
                    elif op == "+":
                        new_item = item + value
                    elif op == "*":
                        new_item = item - value

                if divide_by_3:
                    new_item = int(new_item / 3)

                if new_item % monkey["test"] == 0:
                    monkeys[monkey["true_index"]]["items"].append(new_item % max_modulo)
                else:
                    monkeys[monkey["false_index"]]["items"].append(
                        new_item % max_modulo
                    )
            monkey["items"] = []

    a, b, *_ = sorted(num_items_per_monkey.values(), key=lambda x: -x)
    return a * b


if __name__ == "__main__":
    for file_name in ("example.txt", "input.txt"):
        with open(file_name) as f:
            lines = [l.rstrip("\n") for l in f.readlines()]

        monkeys = parse_monkeys(lines)
        part_1_answer = process_monkeys(monkeys, num_rounds=20, divide_by_3=True)

        monkeys = parse_monkeys(lines)
        part_2_answer = process_monkeys(monkeys, num_rounds=10000, divide_by_3=False)

        print(f"For {file_name}:")
        print(f"Part 1 answer: {part_1_answer}")
        print(f"Part 2 answer: {part_2_answer}")

File: 2022/examples/puzzle_11/main.rs
use parse_display::{Display, FromStr};
use std::collections::HashMap;
use std::time::Instant;

#[derive(Clone, Copy, Debug, Display, FromStr)]

enum Operation {
    #[display("Operation: new = old * old")]
    SQUARE,
    #[display("Operation: new = old + {0}")]
    ADD(u128),
    #[display("Operation: new = old * {0}")]
    MULTIPLY(u128),
}

impl Operation {
    fn apply(&self, item: u128, divide_by_factor: u128) -> u128 {
        let new_item = match self {
            Operation::SQUARE => item * item,
            Operation::ADD(value) => item + value,
            Operation::MULTIPLY(value) => item * value,
        };
        new_item / divide_by_factor
    }
}

#[derive(Debug)]
struct Monkey {
    operation: Operation,
    test_modulo: u128,
    test_true: usize,
    test_false: usize,
}

impl Monkey {
    fn apply_test(&self, item: u128) -> usize {
        if item % self.test_modulo == 0 {
            self.test_true
        } else {
            self.test_false
        }
    }
}

#[derive(FromStr, Debug, PartialEq)]
#[display("  Starting items: {items}")]
#[from_str(new = Self::new(items))]
struct Items {
    items: Vec<u128>,
}

impl Items {
    fn new(s: String) -> Self {
        Items {
            items: s
                .split(", ")
                .collect::<Vec<&str>>()
                .iter()
                .map(|x| x.parse::<u128>().unwrap())
                .collect(),
        }
    }

    fn len(&self) -> usize {
        self.items.len()
    }

    fn pop(&mut self) -> Option<u128> {
        self.items.pop()
    }

    fn push(&mut self, item: u128) {
        self.items.push(item);
    }
}

fn parse_last<T: std::str::FromStr>(line: &str) -> T
where
    <T as std::str::FromStr>::Err: std::fmt::Debug,
{
    let components = line.split_whitespace().collect::<Vec<&str>>();
    components.last().unwrap().parse::<T>().unwrap()
}

fn parse_monkeys(contents: &str) -> (HashMap<usize, Monkey>, HashMap<usize, Items>) {
    let mut monkey_items: HashMap<usize, Items> = HashMap::new();
    let mut monkeys: HashMap<usize, Monkey> = HashMap::new();

    // properties of a monkey that we'll update as we parse
    let mut monkey_index: usize = 0;
    let mut operation: Operation = Operation::SQUARE;
    let mut test_modulo: u128 = 0;
    let mut test_true: usize = 0;
    let mut test_false: usize = 0;

    let mut lines = contents.lines().peekable();
    while let Some(line) = lines.next() {
        if line.contains("Monkey") {
            monkey_index = parse_last(&line.replace(":", ""));
        } else if line.contains("Starting") {
            let items = String::from(line).parse().unwrap();
            monkey_items.insert(monkey_index, items);
        } else if line.contains("Operation") {
            operation = line.trim().parse().unwrap();
        } else if line.contains("Test") {
            test_modulo = parse_last(line);
        } else if line.contains("true") {
            test_true = parse_last(line);
        } else if line.contains("false") {
            test_false = parse_last(line);
        }
        if lines.peek().is_none() || line.is_empty() {
            let monkey = Monkey {
                operation,
                test_modulo,
                test_true,
                test_false,
            };
            monkeys.insert(monkey_index, monkey);
        }
    }
    (monkeys, monkey_items)
}

fn process_monkeys(contents: &str, num_rounds: u32, divide_by_factor: u128) -> usize {
    let (mut monkeys, mut monkey_items) = parse_monkeys(contents);
    let mut num_observed: HashMap<usize, usize> = HashMap::new();
    let mut move_queue: Vec<(usize, u128)> = Vec::new();
    let num_monkeys = monkey_items.len();

    // modulus by this to prevent overflows
    let max_modulus: u128 = monkeys.iter().map(|(_, x)| x.test_modulo).product();

    for _ in 0..num_rounds {
        for monkey_index in 0..num_monkeys {
            // do any moves from previous monkey
            for (move_monkey_index, item) in move_queue.iter() {
                match monkey_items.get_mut(move_monkey_index) {
                    Some(items) => {
                        items.push(*item);
                    }
                    _ => {}
                }
            }
            move_queue = Vec::new();

            let monkey = monkeys.get_mut(&monkey_index).unwrap();
            let num_items = monkey_items[&monkey_index].len();

            // update the number observed
            num_observed
                .entry(monkey_index)
                .and_modify(|e| *e += num_items)
                .or_insert(num_items);

            for _ in 0..monkey_items[&monkey_index].len() {
                match monkey_items.get_mut(&monkey_index) {
                    Some(items) => {
                        let item = items.pop().unwrap();
                        let new_item = monkey.operation.apply(item, divide_by_factor) % max_modulus;
                        let new_monkey_index = monkey.apply_test(new_item);
                        move_queue.push((new_monkey_index, new_item));
                    }
                    _ => {}
                }
            }
        }
    }
    let mut items_seen = num_observed.values().collect::<Vec<&usize>>();
    items_seen.sort_by(|a, b| b.cmp(a));
    items_seen[0] * items_seen[1]
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::*;

    #[test]
    fn test_part_1_example() {
        assert_eq!(process_monkeys(include_str!("./example.txt"), 20, 3), 10605);
    }

    #[test]
    fn test_part_2_example() {
        assert_eq!(
            process_monkeys(include_str!("./example.txt"), 10_000, 1),
            2713310158
        );
    }
}

fn main() {
    let start = Instant::now();
    let contents = include_str!("./input.txt");
    let part_1_answer = process_monkeys(contents, 20, 3);
    println!("Answer for part 1 is: {}", part_1_answer);
    let part_2_answer = process_monkeys(contents, 10_000, 1);
    println!("Answer for part 2 is: {}", part_2_answer);
    let duration = start.elapsed();
    println!("Took {:?} to solve this puzzle", duration);
}
